src/main/java/uk/sky/cirrus/locking/CassandraLockingMechanism.java

```diff
@@ -0,0 +1,126 @@
+
+        } catch (DriverException e) {
+            throw new CannotAcquireLockException("Query to prepare locks queries failed", e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Returns true if successfully inserted lock.
+     * Returns false if current lock is owned by this client.
+     * Returns false if WriteTimeoutException thrown.
+     *
+     * @throws CannotAcquireLockException if any DriverException thrown while executing queries.
+     */
+    @Override
+    public boolean acquire() throws CannotAcquireLockException {
+        try {
+            ResultSet resultSet = session.execute(insertLockQuery.bind(lockName, clientId));
+            Row currentLock = resultSet.one();
+            if (currentLock.getBool("[applied]") || clientId.equals(currentLock.getString("client"))) {
+                return true;
+            } else {
+                log.info("Lock currently held by {}", currentLock);
+                return false;
+            }
+        } catch (WriteTimeoutException wte) {
+            log.warn("Query to acquire lock for {} failed to execute: {}", clientId, wte.getMessage());
+            return false;
+        } catch (DriverException de) {
+            throw new CannotAcquireLockException(String.format("Query to acquire lock %s for client %s failed to execute", lockName, clientId), de);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Deletes lock from locks table.
+     * Will retry to release lock if WriteTimeoutException thrown.
+     * If a WriteTimeoutException has previously been thrown this
+     * will check if the lock did actually successfully deleted.
+     *
+     * @throws CannotAcquireLockException if any DriverException thrown while executing queries.
+     */
+    @Override
+    public void release() throws CannotReleaseLockException {
+        while (true) {
+            try {
+                ResultSet resultSet = session.execute(deleteLockQuery.bind(lockName, clientId));
+                Row result = resultSet.one();
+
+                if (result.getBool("[applied]") || !result.getColumnDefinitions().contains("client")) {
+                    log.info("Lock released for {} by client {} at: {}", lockName, clientId, System.currentTimeMillis());
+                    return;
+                }
+
+                String clientReleasingLock = result.getString("client");
+                if (!clientReleasingLock.equals(clientId)) {
+                    if (isRetryAfterWriteTimeout) {
+                        log.info("Released lock for client {} in retry attempt after WriteTimeoutException", clientReleasingLock);
+                        return;
+                    } else {
+                        throw new CannotReleaseLockException(
+                                String.format("Lock %s attempted to be released by a non lock holder (%s). Current lock holder: %s", lockName, clientId, clientReleasingLock));
+                    }
+                }
+
```

Comment:
I feel like we may be missing a case here which would lead to an infinite loop.

As far as I can tell, we only ever exit if
1. Applied returns true
2. Was previously deleted
3. Client ID doesn't match

If applied continually returns false for some reason it seems like we'll just retry forever. I can't immediately see how that would ever happen, but shouldn't we guard against that case? I think having the retry mechanism at all here is a bit dodgy and complicates things, but if we want it let's try to do it so we can 100% prevent any loops (like with an outer check for max timeout which then will always ensure we break after a while).